tests/typescript-requiring-type-checking/index.ts

```diff
@@ -10,4 +10,10 @@ export class NI {
     public makeAwesomer(): void {
         this._awesomeLevel += 1;
     }
+
+    public async asyncAwesomeness(): Promise<void> {
+        await new Promise(resolve => {
+            resolve(true);
```

Comment:
It's interesting that we don't end up catching this, but may be an artifact of direct Promise usage being harder to type check. But the types don't match, the function signature says `Promise<void>` but it's actually `Promise<boolean>`.

In practice if this was needed for some reason we should just do:

```ts
public async asyncAwesomeness(): Promise<boolean> {
    return await true;
}
```

To minimize ever using `Promise` directly and prefer `async-await`

Or alternatively (but less desirable):

```ts
public async asyncAwesomeness(): Promise<boolean> {
    return Promise.resolve(true);
}
```

Hopefully both of which are better typed.

Reply:
Actually I'm surprised either way that this isn't a type error. Maybe strict type checking isn't enabled here?

Reply:
This comment revealed many interesting bits:
1. we don't set `strict: true` right now and in fact we set `strictNullChecks: false` because that's the default configuration for our ruleset, at least until we fix #73. 

2. That said, I don't think `strict` has anything to do with type errors like this being caught. Locally I do see a red squiggly if I create a type mismatch, but `npm run test` and `npm run lint` both succeed. Nothing in the pipeline ever does a TypeScript compile right now, so type errors are allowed in our tests I guess! 😅

3. The return type of this snippet (and the original more convoluted snippet) is correctly `Promise<void>` since it doesn't return the result of the await. It's a bit subtle, but it's allowed by our rules. Do you see any problem with that? I'm not sure of a way to block it.
```ts
public async asyncAwesomeness(): Promise<void> {
    await Promise.resolve(true);
}
```

4. I hadn't fully understood the [configuration options for @typescript-eslint/return-await](https://typescript-eslint.io/rules/return-await/#options) until now. Do we have a preference between these? According to the table in the rule docs, when outside a try-catch they are functionally equivalent and it's just a style difference. The current decision recorded in #115 is to stick with the default config, `in-try-catch`

```ts
    // allowed by 'always', blocked by 'in-try-catch'
    public async asyncAwesomenessA(): Promise<boolean> {
        return await Promise.resolve(true);
    }

    // blocked by 'always', allowed by 'in-try-catch'
    public async asyncAwesomenessB(): Promise<boolean> {
        return Promise.resolve(true);
    }
```

![image](https://github.com/user-attachments/assets/c85866c0-e302-4771-8d70-8c561abe1f38)

Reply:
I reset myself to look at it again. Both those examples seem convoluted under the assumption that we should not mix `async-await` style code and `Promise()/then()` style code in the same function. Neither should be encouraged and that particular example should not be `async`:

```ts
public asyncAwesomenessC(): Promise<boolean> {
    return Promise.resolve(true);
}
```

But I'd need to read through the error handling parts again to remember the context

Reply:
Looking at it again, I think I'd prefer the `always` configuration. Any Promise returned by an `async` function should be awaited. If you are making a helper, etc. that directly deals with `Promise` instances then you could choose to not mark the function async.

```ts
function timeout(timeMs: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, timeMs));
}
```

If you end up in a situation where you mix and match `async` functions and `Promise` handling:
1. Don't, betting you can refactor the Promise creation into a separate function
2. Even if not, all promises will be awaited making the return paths easier to reason about

```ts
async function timeout(timeMs: number): Promise<void> {
    return await new Promise(resolve => setTimeout(resolve, timeMs));
}
```

I'm gonna try that change in nimble and see what happens

Reply:
This is the resulting PR in nimble: https://github.com/ni/nimble/pull/2412

I think it looks like an improvement and the fixer worked well. Fixer results needed a couple of tweaks to make some inferred generics explicit (comment in PR) but was a straightforward change.

Only difference is this thought:

> If you are making a helper, etc. that directly deals with Promise instances then you could choose to not mark the function async.

Was incorrect. We have [promise-function-async](https://typescript-eslint.io/rules/promise-function-async/) enabled which forces any function that returns a Promise to be async. But I actually like how that combination works. Every promise awaited and every function that returns a Promise is async.

Reply:
Thanks for trying it out! In addition to the reasons you mentioned, I like that `always` doesn't require different code in try-catch vs outside try-catch.

I updated the rule definition and made the test case pass. @rajsite once you approve I think this PR is good to complete!